src/hotspot/cpu/x86/jvmciCodeInstaller_x86.cpp

```diff
@@ -172,23 +173,26 @@ void CodeInstaller::pd_relocate_JavaMethod(CodeBuffer &, JVMCIObject hotspot_met
     case INVOKESTATIC: {
       assert(method == NULL || method->is_static(), "cannot call non-static method with invokestatic");
 
-      NativeCall* call = nativeCall_at(_instructions->start() + pc_offset);
+      call = nativeCall_at(_instructions->start() + pc_offset);
       call->set_destination(SharedRuntime::get_resolve_static_call_stub());
       _instructions->relocate(call->instruction_address(),
                                              relocInfo::static_call_type, Assembler::call32_operand);
       break;
     }
     case INVOKESPECIAL: {
       assert(method == NULL || !method->is_static(), "cannot call static method with invokespecial");
-      NativeCall* call = nativeCall_at(_instructions->start() + pc_offset);
+      call = nativeCall_at(_instructions->start() + pc_offset);
       call->set_destination(SharedRuntime::get_resolve_opt_virtual_call_stub());
       _instructions->relocate(call->instruction_address(),
                               relocInfo::opt_virtual_call_type, Assembler::call32_operand);
       break;
     }
     default:
       JVMCI_ERROR("invalid _next_call_type value");
-      break;
+      return;
+  }
+  if (os::is_MP() && !call->is_displacement_aligned()) {
```

Comment:
You are checking for `MP` in current era? Why not always require alignment?

Reply:
I agree.  I think C2 has always aligned and C1 used to check is_MP but no longer does.  Requiring JVMCI compilers to always align seems right, particularly since we don't expose is `is_MP` through JVMCI.  Graal and C1 actually appear to over align the displacement by aligning it to BytesPerWord while C2 always aligns to 4 which is all that is required.  It's odd that alignment check in NativeCall isn't simply `displacement_offset() % 4`.  The existing check implies that it's ok to use a misaligned offset as long as it starts and ends within an 8 byte region but I don't know that that would really work and none of the compilers actually take advantage of it.  That's probably beyond the scope of this PR.

Reply:
I will remove the `is_MP()` calls.

>while C2 always aligns to 4

I'm having trouble finding where that is done - can you please point it out.

Reply:
It's deeply hidden.  in x86_64.ad, it specifies the call alignment as [4](https://github.com/tkrodriguez/jdk/blob/master/src/hotspot/cpu/x86/x86_64.ad#L13028) and in [compute_padding](https://github.com/tkrodriguez/jdk/blob/master/src/hotspot/cpu/x86/x86_64.ad#L493) it uses that alignment value to align the offset of the displacement.

Reply:
Should we fix C1 and JVMCI to do the same as C2? As separate RFE.

Reply:
There's nothing to fix on the JVMCI side other than ensuring that we're asserting the correct alignment restrictions.  Reducing the alignment to 4 instead of 8 would be a pure Graal change.

Reply:
> It's deeply hidden

Thanks - doubt I would ever have found that in reasonable time.

Reply:
I meant C1 and Graal. Yes, in JVMCI we should just check correct alignment.